### ISSUE TITLE
Hosted documentation of sway diagnostic codes

### ISSUE THREAD

**Issue by sdankel** [FuelLabs/sway#4990]
As part of Expressive Diagnostics, we will have a unique code for each error/warning type.

https://github.com/FuelLabs/sway-rfcs/pull/30#issuecomment-1684253622

We should create something similar to Rust's error index to provide more detailed documentation of each error/warning code.

https://doc.rust-lang.org/error_codes/error-index.html

**Comment by sdankel** [comment]
@eureka-cpu It would be great if you could share your thoughts on how the error should be defined to make doc generation easy.

**Comment by ironcev** [comment]
Duplicate of #3512. I'll close that one since this issue is already added to the Sway Tooling project.

These will perhaps be additional issues, but just to get the whole picture, motivated by Rust:
- `forc explain E01001` would be cool to have
- as explained in the [Future Possibilities chapter in RFC](https://github.com/FuelLabs/sway-rfcs/blob/048419a5c314dd6d8f1cc9828cf027ebab97437e/rfcs/0000-expressive-diagnostics.md#future-possibilities) diagnostics should also link to docs one day
  - having CI checking the links, even chapter names would then be a must

I like how our docs.fuel.network space is evolving and looking forward to integrate it more closely with the compiler :-)

**Comment by eureka-cpu** [comment]
> @eureka-cpu It would be great if you could share your thoughts on how the error should be defined to make doc generation easy.

Since the errors will be written in Rust, we could just produce the documentation for that specific module with `cargo doc`. If we're worried about how it looks, we could use GH pages to deploy a version of it with modified CSS so it can match the look and feel of the already established doc hub. cc @sarahschwartz

**Comment by ironcev** [comment]
> Since the errors will be written in Rust, we could just produce the documentation for that specific module with cargo doc.

I like the idea of having the diagnostic docs in the code:
- the closer the docs are to the place where the diagnostic is defined the easier to developers to immediately write them :smile: 
- the presence and the structure of the docs can be checked at the compile time within the diagnostics proc-macro
- we get the `forc explain` almost for free because the docs are already embedded within the diagnostics
- Clippy does it exactly the same :wink:

The "just" part will very likely require some effort aside from modifying CSS but it should also be a moderate effort. The pure Rust module documentation will contain additional compiler specific details and the enum variants for errors. We will need to extract only individual docs, place them separately, and connect them with the index. Should be straightforward.

I'll then assume that the docs are in code, create first examples, and support it in the proc-macro. Let's catch up and review the approach once the first examples are there.

**Comment by eureka-cpu** [comment]
Yep, all good points. And if you need any insight to `cargo doc` internals, you might take a look at `forc doc` since it is structurally based on it. You shouldn't have any problem generating error code docs and then managing the output with GitHub actions.
